File: pipeline/reimpl/cocrdlic.py
"""
Reimplementation of COCRDLIC — CardDemo Credit Card List Screen.

CICS online program. Lists credit cards (7 per page) with filtering options:
  - Admin users: can see all cards or filter by account/card number
  - Regular users: see only cards for their own account

Selection codes:
  S = View card detail → COCRDSLC
  U = Update card       → COCRDUPC

Navigation: PF3=exit, PF7=prev page, PF8=next page
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Optional

from .carddemo_data import (
    CarddemoCommarea, ScreenHeader, CardRecord,
    CCDA_MSG_INVALID_KEY, DFHENTER, DFHPF3, DFHPF7, DFHPF8,
)


WS_PGMNAME = "COCRDLIC"
WS_TRANID  = "CCLI"
PAGE_SIZE  = 7


@dataclass
class CardListRow:
    card_num: str = ""
    acct_id: int = 0
    embossed_name: str = ""
    expiration_date: str = ""
    active_status: str = ""
    sel_flag: str = ""


@dataclass
class CardListResult:
    rows: list[CardListRow] = field(default_factory=list)
    page_num: int = 1
    has_next_page: bool = False
    has_prev_page: bool = False
    xctl_program: str = ""
    selected_card_num: str = ""
    selected_action: str = ""
    commarea: Optional[CarddemoCommarea] = None
    screen: Optional[ScreenHeader] = None
    message: str = ""
    error: bool = False
    return_to_prev: bool = False


class CardRepository:
    def __init__(self, cards: list[CardRecord]):
        self._all = sorted(cards, key=lambda c: c.card_num)

    def get_page(
        self,
        page_num: int,
        acct_filter: int = 0,
        card_filter: str = "",
    ) -> tuple[list[CardRecord], bool]:
        cards = self._all
        if acct_filter:
            cards = [c for c in cards if c.card_acct_id == acct_filter]
        if card_filter.strip():
            cards = [c for c in cards if c.card_num.startswith(card_filter.strip())]
        start = (page_num - 1) * PAGE_SIZE
        end = start + PAGE_SIZE
        return cards[start:end], len(cards) > end


def process_card_list(
    eibcalen: int,
    eibaid: str,
    commarea: CarddemoCommarea,
    card_repo: CardRepository,
    page_num: int = 1,
    acct_filter: int = 0,
    card_filter: str = "",
    selected_rows: list[tuple[str, str]] = None,
) -> CardListResult:
    """Process credit card list — mirrors COCRDLIC PROCEDURE DIVISION."""
    result = CardListResult()

    if eibcalen == 0:
        result.return_to_prev = True
        result.xctl_program = "COSGN00C"
        return result

    result.commarea = commarea

    if commarea.cdemo_pgm_context == 0:
        commarea.cdemo_pgm_context = 1
        return _load_page(page_num, acct_filter, card_filter, card_repo, commarea, result)

    if eibaid in (DFHENTER, "ENTER"):
        return _process_enter(
            selected_rows or [], page_num, acct_filter, card_filter,
            card_repo, commarea, result,
        )
    elif eibaid in (DFHPF3, "PF3"):
        back = commarea.cdemo_from_program or "COMEN01C"
        result.xctl_program = back
        result.return_to_prev = True
        result.commarea = commarea
        return result
    elif eibaid in (DFHPF7, "PF7"):
        return _load_page(max(1, page_num - 1), acct_filter, card_filter, card_repo, commarea, result)
    elif eibaid in (DFHPF8, "PF8"):
        return _load_page(page_num + 1, acct_filter, card_filter, card_repo, commarea, result)
    else:
        result.error = True
        result.message = CCDA_MSG_INVALID_KEY
        return _load_page(page_num, acct_filter, card_filter, card_repo, commarea, result)


def _load_page(page_num, acct_filter, card_filter, repo, commarea, result):
    cards, has_next = repo.get_page(page_num, acct_filter, card_filter)
    result.page_num = page_num
    result.has_next_page = has_next
    result.has_prev_page = page_num > 1
    result.rows = [
        CardListRow(
            card_num=c.card_num,
            acct_id=c.card_acct_id,
            embossed_name=c.card_embossed_name.strip(),
            expiration_date=c.card_expiration_date,
            active_status=c.card_active_status,
        )
        for c in cards
    ]
    if not cards:
        result.error = True
        result.message = "NO RECORDS FOUND FOR THIS SEARCH CONDITION."
    else:
        result.message = "TYPE S FOR DETAIL, U TO UPDATE ANY RECORD"
    result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
    result.screen.errmsg = result.message
    result.commarea = commarea
    return result


def _process_enter(selected_rows, page_num, acct_filter, card_filter, repo, commarea, result):
    selections = [(f, c) for f, c in selected_rows if f and f.strip() not in ("", "\x00") and c.strip()]
    if len(selections) > 1:
        result.error = True
        result.message = "PLEASE SELECT ONLY ONE RECORD TO VIEW OR UPDATE"
        return _load_page(page_num, acct_filter, card_filter, repo, commarea, result)

    if selections:
        action, card_num = selections[0]
        a = action.strip().upper()
        if a not in ("S", "U"):
            result.error = True
            result.message = "INVALID ACTION CODE"
            return _load_page(page_num, acct_filter, card_filter, repo, commarea, result)

        result.selected_card_num = card_num.strip()
        result.selected_action = a
        commarea.cdemo_pgm_context = 0

        if a == "S":
            result.xctl_program = "COCRDSLC"
        else:
            result.xctl_program = "COCRDUPC"
        result.commarea = commarea
        return result

    return _load_page(page_num, acct_filter, card_filter, repo, commarea, result)


# ── Differential harness runner adapter ──────────────────────────────────────

_SEED_CARDS = [
    CardRecord(card_num=f"411111111111{i:04d}", card_acct_id=10000001, card_cvv_cd=100+i,
               card_embossed_name=f"TEST USER {i:04d}", card_expiration_date="2028-12-31",
               card_active_status="Y")
    for i in range(1, 10)
]


def run_vector(inputs: dict) -> dict:
    """Canonical runner entry point for the differential harness.

    SCENARIO selects a hardcoded test path:
      FIRST_ENTRY   — context=0, first entry → load page 1
      LIST_PAGE_1   — display page 1
      SELECT_VIEW   — select card for view → COCRDSLC
      SELECT_UPDATE — select card for update → COCRDUPC
      PF3_RETURN    — press PF3 → return to menu
      INVALID_KEY   — press PF9 → invalid key error
    """
    scenario = inputs.get("SCENARIO", "FIRST_ENTRY")

    repo = CardRepository(list(_SEED_CARDS))

    commarea = CarddemoCommarea(
        cdemo_from_tranid="CM00",
        cdemo_from_program="COMEN01C",
        cdemo_user_id="USER0001",
        cdemo_user_type="U",
        cdemo_pgm_context=1,
    )

    if scenario == "FIRST_ENTRY":
        commarea.cdemo_pgm_context = 0
        result = process_card_list(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            card_repo=repo, page_num=1, selected_rows=[],
        )
    elif scenario == "LIST_PAGE_1":
        result = process_card_list(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            card_repo=repo, page_num=1, selected_rows=[],
        )
    elif scenario == "SELECT_VIEW":
        result = process_card_list(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            card_repo=repo, page_num=1,
            selected_rows=[("S", "4111111111110001")],
        )
    elif scenario == "SELECT_UPDATE":
        result = process_card_list(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            card_repo=repo, page_num=1,
            selected_rows=[("U", "4111111111110002")],
        )
    elif scenario == "PF3_RETURN":
        result = process_card_list(
            eibcalen=100, eibaid="PF3", commarea=commarea,
            card_repo=repo, page_num=1, selected_rows=[],
        )
    elif scenario == "INVALID_KEY":
        result = process_card_list(
            eibcalen=100, eibaid="PF9", commarea=commarea,
            card_repo=repo, page_num=1, selected_rows=[],
        )
    else:
        commarea.cdemo_pgm_context = 0
        result = process_card_list(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            card_repo=repo, page_num=1, selected_rows=[],
        )

    out: dict[str, str] = {
        "PAGE_NUM": str(result.page_num),
        "ROW_COUNT": str(len(result.rows)),
        "HAS_NEXT": "Y" if result.has_next_page else "N",
        "HAS_PREV": "Y" if result.has_prev_page else "N",
        "ERROR": "Y" if result.error else "N",
        "MESSAGE": result.message,
        "XCTL_PROGRAM": result.xctl_program or "",
        "SELECTED_CARD_NUM": result.selected_card_num,
        "SELECTED_ACTION": result.selected_action,
        "RETURN_TO_PREV": "Y" if result.return_to_prev else "N",
    }
    for i, row in enumerate(result.rows):
        out[f"ROW_{i}_CARD_NUM"] = row.card_num
        out[f"ROW_{i}_ACCT_ID"] = str(row.acct_id)
        out[f"ROW_{i}_NAME"] = row.embossed_name
        out[f"ROW_{i}_STATUS"] = row.active_status
    return out

File: pipeline/reimpl/java/runner/src/main/java/com/modernization/masquerade/runner/programs/Cocrdlic.java
package com.modernization.masquerade.runner.programs;

import com.modernization.masquerade.runner.ProgramRunner;

import java.util.*;

/**
 * Java reimplementation of COCRDLIC — CardDemo Credit Card List Screen.
 *
 * <p>Mirrors {@code pipeline/reimpl/cocrdlic.py}. Python is the source of truth.
 *
 * <p>Standalone business method: {@link #processCardList}. The {@link #runVector}
 * method is a thin adapter that builds scenario-specific parameters and delegates.
 */
public class Cocrdlic implements ProgramRunner {

    private static final int PAGE_SIZE = 7;

    // ── Inner data types ─────────────────────────────────────────────────

    static class CardRecord {
        final String cardNum;
        final String acctId;
        final String embossedName;
        final String activeStatus;

        CardRecord(String cardNum, String acctId, String embossedName, String activeStatus) {
            this.cardNum = cardNum;
            this.acctId = acctId;
            this.embossedName = embossedName;
            this.activeStatus = activeStatus;
        }
    }

    static class CardListRow {
        final String cardNum;
        final String acctId;
        final String embossedName;
        final String activeStatus;

        CardListRow(String cardNum, String acctId, String embossedName, String activeStatus) {
            this.cardNum = cardNum;
            this.acctId = acctId;
            this.embossedName = embossedName;
            this.activeStatus = activeStatus;
        }
    }

    static class CardListResult {
        List<CardListRow> rows = new ArrayList<>();
        int pageNum = 1;
        boolean hasNext = false;
        boolean hasPrev = false;
        boolean error = false;
        String message = "";
        String xctlProgram = "";
        String selectedCardNum = "";
        String selectedAction = "";
        boolean returnToPrev = false;
    }

    static class CardRepo {
        private final List<CardRecord> all;

        CardRepo(List<CardRecord> cards) {
            this.all = new ArrayList<>(cards);
            this.all.sort(Comparator.comparing(c -> c.cardNum));
        }

        PageResult getPage(int pageNum, int acctFilter, String cardFilter) {
            List<CardRecord> filtered = all;
            if (acctFilter != 0) {
                List<CardRecord> tmp = new ArrayList<>();
                String af = String.valueOf(acctFilter);
                for (CardRecord c : all) {
                    if (c.acctId.equals(af)) tmp.add(c);
                }
                filtered = tmp;
            }
            if (cardFilter != null && !cardFilter.trim().isEmpty()) {
                String prefix = cardFilter.trim();
                List<CardRecord> tmp = new ArrayList<>();
                for (CardRecord c : filtered) {
                    if (c.cardNum.startsWith(prefix)) tmp.add(c);
                }
                filtered = tmp;
            }
            int start = (pageNum - 1) * PAGE_SIZE;
            int end = Math.min(start + PAGE_SIZE, filtered.size());
            List<CardRecord> page = (start < filtered.size())
                    ? filtered.subList(start, end) : new ArrayList<>();
            return new PageResult(page, filtered.size() > end);
        }
    }

    static class PageResult {
        final List<CardRecord> records;
        final boolean hasNext;
        PageResult(List<CardRecord> records, boolean hasNext) {
            this.records = records;
            this.hasNext = hasNext;
        }
    }

    // ── Standalone business method ──────────────────────────────────────

    /**
     * Process credit card list — mirrors Python's process_card_list.
     *
     * @param eibcalen     CICS EIBCALEN
     * @param eibaid       aid key pressed
     * @param pgmContext   commarea program context (0 = initial)
     * @param fromProgram  commarea cdemo_from_program
     * @param cardRepo     card repository
     * @param pageNum      current page number
     * @param acctFilter   account filter (0 = no filter)
     * @param cardFilter   card number prefix filter
     * @param selectedRows list of (selFlag, cardNum) pairs from screen
     * @return CardListResult with all output fields
     */
    CardListResult processCardList(
            int eibcalen,
            String eibaid,
            int pgmContext,
            String fromProgram,
            CardRepo cardRepo,
            int pageNum,
            int acctFilter,
            String cardFilter,
            List<String[]> selectedRows
    ) {
        CardListResult result = new CardListResult();

        if (eibcalen == 0) {
            result.returnToPrev = true;
            result.xctlProgram = "COSGN00C";
            return result;
        }

        if (pgmContext == 0) {
            return loadPage(pageNum, acctFilter, cardFilter, cardRepo, result);
        }

        switch (eibaid) {
            case "ENTER":
                return processEnter(selectedRows, pageNum, acctFilter, cardFilter, cardRepo, result);
            case "PF3": {
                String back = (fromProgram != null && !fromProgram.isEmpty()) ? fromProgram : "COMEN01C";
                result.xctlProgram = back;
                result.returnToPrev = true;
                return result;
            }
            case "PF7":
                return loadPage(Math.max(1, pageNum - 1), acctFilter, cardFilter, cardRepo, result);
            case "PF8":
                return loadPage(pageNum + 1, acctFilter, cardFilter, cardRepo, result);
            default:
                result.error = true;
                return loadPage(pageNum, acctFilter, cardFilter, cardRepo, result);
        }
    }

    private CardListResult loadPage(int pageNum, int acctFilter, String cardFilter,
                                     CardRepo repo, CardListResult result) {
        PageResult pr = repo.getPage(pageNum, acctFilter, cardFilter);
        result.pageNum = pageNum;
        result.hasNext = pr.hasNext;
        result.hasPrev = pageNum > 1;
        result.rows = new ArrayList<>();
        for (CardRecord c : pr.records) {
            result.rows.add(new CardListRow(
                    c.cardNum, c.acctId, c.embossedName.trim(), c.activeStatus));
        }
        if (pr.records.isEmpty()) {
            result.error = true;
            result.message = "NO RECORDS FOUND FOR THIS SEARCH CONDITION.";
        } else {
            result.message = "TYPE S FOR DETAIL, U TO UPDATE ANY RECORD";
        }
        return result;
    }

    private CardListResult processEnter(
            List<String[]> selectedRows, int pageNum, int acctFilter,
            String cardFilter, CardRepo repo, CardListResult result
    ) {
        // Collect valid selections
        List<String[]> selections = new ArrayList<>();
        if (selectedRows != null) {
            for (String[] sel : selectedRows) {
                String flag = sel[0];
                String cardNum = sel[1];
                if (flag != null && !flag.trim().isEmpty() && !flag.trim().equals("\0")
                        && cardNum != null && !cardNum.trim().isEmpty()) {
                    selections.add(sel);
                }
            }
        }

        if (selections.size() > 1) {
            result.error = true;
            result.message = "PLEASE SELECT ONLY ONE RECORD TO VIEW OR UPDATE";
            return loadPage(pageNum, acctFilter, cardFilter, repo, result);
        }

        if (!selections.isEmpty()) {
            String action = selections.get(0)[0].trim().toUpperCase();
            String cardNum = selections.get(0)[1].trim();

            if (!action.equals("S") && !action.equals("U")) {
                result.error = true;
                result.message = "INVALID ACTION CODE";
                return loadPage(pageNum, acctFilter, cardFilter, repo, result);
            }

            result.selectedCardNum = cardNum;
            result.selectedAction = action;

            if (action.equals("S")) {
                result.xctlProgram = "COCRDSLC";
            } else {
                result.xctlProgram = "COCRDUPC";
            }
            return result;
        }

        return loadPage(pageNum, acctFilter, cardFilter, repo, result);
    }

    // ── Seed data ───────────────────────────────────────────────────────

    private List<CardRecord> buildSeedCards() {
        List<CardRecord> list = new ArrayList<>();
        for (int i = 1; i <= 9; i++) {
            list.add(new CardRecord(
                    String.format("411111111111%04d", i),
                    "10000001",
                    String.format("TEST USER %04d", i),
                    "Y"
            ));
        }
        return list;
    }

    // ── Thin runVector adapter ──────────────────────────────────────────

    @Override
    public Map<String, String> runVector(Map<String, String> inputs) {
        String scenario = inputs.getOrDefault("SCENARIO", "FIRST_ENTRY");

        CardRepo repo = new CardRepo(buildSeedCards());
        List<String[]> noSelection = new ArrayList<>();

        CardListResult result;

        switch (scenario) {
            case "FIRST_ENTRY":
                result = processCardList(100, "ENTER", 0, "COMEN01C",
                        repo, 1, 0, "", noSelection);
                break;

            case "LIST_PAGE_1":
                result = processCardList(100, "ENTER", 1, "COMEN01C",
                        repo, 1, 0, "", noSelection);
                break;

            case "SELECT_VIEW": {
                List<String[]> sel = new ArrayList<>();
                sel.add(new String[]{"S", "4111111111110001"});
                result = processCardList(100, "ENTER", 1, "COMEN01C",
                        repo, 1, 0, "", sel);
                break;
            }

            case "SELECT_UPDATE": {
                List<String[]> sel = new ArrayList<>();
                sel.add(new String[]{"U", "4111111111110002"});
                result = processCardList(100, "ENTER", 1, "COMEN01C",
                        repo, 1, 0, "", sel);
                break;
            }

            case "PF3_RETURN":
                result = processCardList(100, "PF3", 1, "COMEN01C",
                        repo, 1, 0, "", noSelection);
                break;

            case "INVALID_KEY":
                result = processCardList(100, "PF9", 1, "COMEN01C",
                        repo, 1, 0, "", noSelection);
                break;

            default:
                result = processCardList(100, "ENTER", 0, "COMEN01C",
                        repo, 1, 0, "", noSelection);
                break;
        }

        Map<String, String> out = new LinkedHashMap<>();
        out.put("PAGE_NUM", String.valueOf(result.pageNum));
        out.put("ROW_COUNT", String.valueOf(result.rows.size()));
        out.put("HAS_NEXT", result.hasNext ? "Y" : "N");
        out.put("HAS_PREV", result.hasPrev ? "Y" : "N");
        out.put("ERROR", result.error ? "Y" : "N");
        out.put("MESSAGE", result.message);
        out.put("XCTL_PROGRAM", result.xctlProgram);
        out.put("SELECTED_CARD_NUM", result.selectedCardNum);
        out.put("SELECTED_ACTION", result.selectedAction);
        out.put("RETURN_TO_PREV", result.returnToPrev ? "Y" : "N");

        for (int i = 0; i < result.rows.size(); i++) {
            CardListRow row = result.rows.get(i);
            out.put("ROW_" + i + "_CARD_NUM", row.cardNum);
            out.put("ROW_" + i + "_ACCT_ID", row.acctId);
            out.put("ROW_" + i + "_NAME", row.embossedName);
            out.put("ROW_" + i + "_STATUS", row.activeStatus);
        }
        return out;
    }
}
